modules/level_set/src/kernels/LevelSetAdvectionSUPG.C

```diff
@@ -17,19 +17,20 @@ LevelSetAdvectionSUPG::validParams()
   InputParameters params = ADKernelGrad::validParams();
   params.addClassDescription(
       "SUPG stablization term for the advection portion of the level set equation.");
-  params += LevelSetVelocityInterface<ADKernelGrad>::validParams();
+  params.addRequiredCoupledVar("velocity", "Velocity vector variable.");
   return params;
 }
 
 LevelSetAdvectionSUPG::LevelSetAdvectionSUPG(const InputParameters & parameters)
-  : LevelSetVelocityInterface<ADKernelGrad>(parameters)
+  : ADKernelGrad(parameters), _velocity(adCoupledVectorValue("velocity"))
 {
 }
 
 ADRealVectorValue
 LevelSetAdvectionSUPG::precomputeQpResidual()
 {
-  computeQpVelocity();
-  ADReal tau = _current_elem->hmin() / (2 * _velocity.norm());
-  return (tau * _velocity) * (_velocity * _grad_u[_qp]);
+  ADReal tau =
+      _current_elem->hmin() /
+      (2 * (_velocity[_qp] + RealVectorValue(libMesh::TOLERANCE * libMesh::TOLERANCE)).norm());
```

Comment:
I'm guess the intention here is to make sure we don't divide by zero if `_velocity[_qp]` is `[0 0 0]`, if so I would change this line to:
```suggestion
      (2 * _velocity[_qp].norm() + libMesh::TOLERANCE);
```

Reply:
@zachmprince  Not just for the reason you mentioned. The norm() uses sqrt function. The derivative of sqrt function at 0 will become infinite, so we try to avoid that in AD. Does this make sense? Thanks.

Reply:
Ah, thank you for your explanation. I apologize for my ignorance on the AD system. This is fine with me then.

Reply:
No worry. Thanks for your review. I guess you pass your final 'exam' of review training. I can ping you in the future for other reviews. :-)

Reply:
FYI, libMesh vectors also have an is_zero method to test for this case